### DISCUSSION TITLE
best-practices/remote-deployment

### DISCUSSION THREAD

**Discussion by giscus** [ryan4yin/nixos-and-flakes-book#143] [Announcements]
# best-practices/remote-deployment

An unofficial and opinionated book for beginners

https://nixos-and-flakes.thiscute.world/best-practices/remote-deployment

**Comment by Rizqirazkafi** [comment]
I build using nixos-rebuild and set the --build-host the same as --remote-host. Does my configuration file e.g : configuration.nix, flake.nix , etc ... got copied to the --build-host? If so, where can I find it?

**Comment by fl42v** [comment]
When using the `nixos-rebiuld` route with `--build-host localhost`, it seems like one also needs `services.openssh` configured on said localhost. "Weird flex but okay"

**Reply by fl42v** [reply]
So, after reading the nixos-rebuild's manpage, the resulting command looked something like that: `nixos-rebuild switch --flake <path_to_flake>#<remote_hostname> --target-host <remote_machine's_ip> --use-remote-sudo --verbose`, plus I had to add my username to `nix.settings.trusted-users` on the remote machine.

**Reply by Propfend** [reply]
The remote host is agentic in the end, even if its minimally. That is something i don't like when automating deployments, most NixOS machines in cloud are not so good and most approaches of deploying NixOS don't even leave the configuration reproducible is host like `nixos-anywhere`. Sometimes the target machine doesn't even have NixOS running. But this is the simplest approach, if you don't need to deploy some machine from scratch then you are good.
